Mem width fix

nathanielnrn opened the pull request:
Creating a draft PR to get feedback from @rachitnigam and others.

The first part of the fix (up to commit "replace mapping of .clone() to cloned()")  creates multiple bram modules based on the size of memories in a calyx program.

Had to wrangle a lot with [this](https://github.com/cucapra/calyx/blob/f33f0a8677cb35cf9c0218561211a9add863ef71/src/backend/xilinx/toplevel.rs#L108-L118) helper function (copied below) due to rust typing and ownership things. Still not completely happy with how this came out.

```
fn external_memories_cells(
    comp: &ir::Component,
) -> Vec<std::rc::Rc<std::cell::RefCell<ir::Cell>>> {
    comp.cells
        .iter()
        .filter(|cell_ref| {
            matches!(cell_ref.borrow().get_attribute("external"), Some(&1))
        })
        .map(|cell_ref| cell_ref.clone())
        .collect()
}
```

Would particularly appreciate feedback on:
1. The [creation of separate modules for each memory](https://github.com/cucapra/calyx/blob/f33f0a8677cb35cf9c0218561211a9add863ef71/src/backend/xilinx/toplevel.rs#L57-L64), specifically, should we look into sharing identically sized memories/doing this some other way?
2. Creating the separate modules by iteration through [these](https://github.com/cucapra/calyx/blob/f33f0a8677cb35cf9c0218561211a9add863ef71/src/backend/xilinx/toplevel.rs#L49-L51) vectors
3. If the helper function can be simplified and improved

Of course any other feedback is always appreciated

rachitnigam commented on src/backend/xilinx/memory_axi.rs:
```diff
@@ -315,9 +331,9 @@ fn bram_logic(
 
     // bram address logic
-    let copy_address = v::Expr::new_slice("copy_addr_offset", 4, 0);
+    let copy_address = v::Expr::new_slice("copy_addr_offset", 4, 0); // does this need to use parameterization?
```
When you have question inside a comment, write something like:
```
XXX(nathanielnrn): ...
```
Same with TODOs:
```
TODO(username): ...
```

rachitnigam commented on src/backend/xilinx/toplevel.rs:
```diff
@@ -79,6 +104,44 @@ impl Backend for XilinxInterfaceBackend {
+    comp.cells
+        .iter()
+        .filter(|cell_ref| {
+            matches!(cell_ref.borrow().get_attribute("external"), Some(&1))
```
You can use `has_attribute` instead of the `match!(..)` call

rachitnigam commented on src/backend/xilinx/toplevel.rs:
```diff
@@ -79,6 +104,44 @@ impl Backend for XilinxInterfaceBackend {
+//should this be a macro?
+fn external_memories_cells(
+    comp: &ir::Component,
+) -> Vec<std::rc::Rc<std::cell::RefCell<ir::Cell>>> {
```
Use `RRC<...>` instead of `Rc<RefCell<...>`

rachitnigam commented on src/backend/xilinx/toplevel.rs:
```diff
@@ -79,6 +104,44 @@ impl Backend for XilinxInterfaceBackend {
+        .collect()
+}
+
+//returns vector of memory widths from a component
```
You can combine these these functions into one that returns a `(WIDTH, SIZE, IDX_SIZE)` tuple for each memory instead of doing this separately in three functions

rachitnigam commented:
To answer your questions:
1. No. Let's get the simplest working thing done first so we can start testing it. I'd like us to add a new `runt` test that checks the generated output of the AXI generator once we are done with #1071.
2. The link doesn't work for me so I'm not sure which vectors
3. Left some inline comments in the code for this

rachitnigam commented on src/backend/xilinx/memory_axi.rs:
```diff
@@ -293,18 +305,22 @@ fn module_mode_fsm(module: &mut v::Module) -> fsm::LinearFsm {
 }
 
 fn bram_logic(
+    index: u64,
```
I'd prefer this to be the name instead of an instance. That way, it's the responsibility of the caller to ensure the name is unique

nathanielnrn commented:
Addressed all of the comments, got rid of almost all of the hardcoded values except the ones marked with xxxs or todos. If @rachitnigam you happen to know how they should behave or where they are from at a glance I will fix these otherwise I will look at them more tomorrow. Think it's ready to be reviewed and hopefully merged so moving it out of draft status.

rachitnigam commented on src/backend/xilinx/memory_axi.rs:
```diff
@@ -162,13 +165,22 @@ impl MemoryInterface for AxiInterface {
 
         // bram reading / writing logic
-        bram_logic(&axi4, &mut module, &mode_fsm, "read_txn_count".into());
+        // TODO(nathanielnrn) fix this
```
Fix this how? In general TODO comments should attempt to explain what is going on in enough detail that someone can understand it 6 months from now. Either extend the comment or point to an issue

rachitnigam commented on src/backend/xilinx/memory_axi.rs:
```diff
@@ -349,7 +369,7 @@ fn bram_logic(
         &axi4.read_data.get("DATA"),
         v::Expr::new_mul(txn_count, 32),
-        32, /* bram data width */
+        data_width.try_into().unwrap(), /* bram data width */
```
Why do you need the try_into call here?

rachitnigam commented on src/backend/xilinx/toplevel.rs:
```diff
@@ -79,13 +98,39 @@ impl Backend for XilinxInterfaceBackend {
-fn external_memories(comp: &ir::Component) -> Vec<String> {
-    // find external memories
+// XXX(nathanielnrn) I __think__ this should always be in the same order because
+// of IdList being in the same order
```
It will be the same order.

rachitnigam commented on src/backend/xilinx/toplevel.rs:
```diff
@@ -79,13 +98,39 @@ impl Backend for XilinxInterfaceBackend {
-    // find external memories
+// XXX(nathanielnrn) I __think__ this should always be in the same order because
+// of IdList being in the same order
+// also, should this be a macro?
```
What should be a macro? This function? Is so why?

nathanielnrn replied:
Initially I was wondering if the helper function should be a macro due to some trouble I was having implementing it. But if the helper function looks ok:
```
fn external_memories_cells(
    comp: &ir::Component,
) -> Vec<calyx::ir::RRC<ir::Cell>> {
    comp.cells
        .iter()
        .filter(|cell_ref| {
            matches!(cell_ref.borrow().get_attribute("external"), Some(&1))
        // find external memories
        .filter(|cell_ref| cell_ref.borrow().attributes.has("external"))
        .cloned()
        .collect()
}
```
And the way it is used in these two functions is fine:

```
// Returns a vector of tuples containing external memory info of form:
// (WIDTH, SIZE, IDX_SIZE)
fn get_mem_info(comp: &ir::Component) -> Vec<(u64, u64, u64)> {
    external_memories_cells(comp)
        .iter()
        .map(|cell_ref| {
            (
                cell_ref.borrow().get_parameter("WIDTH").unwrap(),
                cell_ref.borrow().get_parameter("SIZE").unwrap(),
                cell_ref.borrow().get_parameter("IDX_SIZE").unwrap(),
            )
        })
        .collect()
}

// Returns Vec<String> of memory names
fn external_memories(comp: &ir::Component) -> Vec<String> {
    external_memories_cells(comp)
        .iter()
        .map(|cell_ref| cell_ref.borrow().name().to_string())
        .collect()
}
```

Then it should be fine to leave it as is.

rachitnigam reviewed (approved):
A few comments here and there but overall looks good. Regarding correctness, we’ll only know if it works when you run it with hw_emu or on an FPGA

nathanielnrn commented:
If you're happy with @rachitnigam  [this](https://github.com/cucapra/calyx/pull/1096#discussion_r921419094) then I will merge

rachitnigam commented:
Yeah, do it!

sampsyo commented on src/backend/xilinx/memory_axi.rs:
```diff
@@ -98,11 +100,11 @@ impl MemoryInterface for AxiInterface {
     ) -> v::Module {
         let mut module = v::Module::new(name);
-        let memory_size = 32;
         let memory_size_bits: u64 = utils::math::bits_needed_for(memory_size); // TODO make memory size parametric
```
Hopefully `memory_size_bits == addr_width`… maybe adding an `assert` here to check that would be cool?

nathanielnrn replied:
This is actually slightly changed in #1103.
One thing that occurred to me on the subject is that because calyx-memory has the width of it's address port defined through an argument, there could be a (perhaps strange? Not sure how common this would be) case where there is a wider address port than needed for the size of a memory. In that case the two wouldn't be equal.

sampsyo commented on src/backend/xilinx/memory_axi.rs:
```diff
@@ -293,18 +306,23 @@ fn module_mode_fsm(module: &mut v::Module) -> fsm::LinearFsm {
     module.add_decl(v::Decl::new_wire("bram_done", 1));
-
-    let mut ram_instance = v::Instance::new("bram", "SINGLE_PORT_BRAM");
+    let suffix_idx = "Memory_controller_axi_".len();
+    let suffix = &name[suffix_idx..];
```
Not a big deal, but it could be ever so slightly nicer to take `suffix` in specifically as a parameter?

sampsyo reviewed (commented):
Super nice work, y'all. Very cool to see the memory parameters extracted and then propagated all the way through the AXI generation. Just wanted to read through so I understand how this all works; you can probably just ignore my two scant comments.
